pxd.c

```diff
@@ -1623,19 +1623,24 @@ static void pxd_finish_remove(struct work_struct *work)
 
 		mutex_unlock(&pxd_dev->disk->queue->sysfs_lock);
 #else
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,13,0)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,15,25)
 	// del_gendisk will try to fsync device
-	// so freeze queue and then mark queue dead to ensure no new reqs
+	// so freeze queue and then *mark queue dead* to ensure no new reqs
 	// gets accepted.
     blk_freeze_queue_start(pxd_dev->disk->queue);
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,15,25)
     blk_mark_disk_dead(pxd_dev->disk);
-#else
-    blk_set_queue_dying(pxd_dev->disk->queue);
-#endif
-#else
+#elif LINUX_VERSION_CODE < KERNEL_VERSION(5,13,0)
+	// del_gendisk will not submit any new IO.
+	// so freeze queue and then queue dying, to ensure no new reqs
+	// gets accepted.
+    blk_freeze_queue_start(pxd_dev->disk->queue);
     blk_set_queue_dying(pxd_dev->disk->queue);
 #endif
+	// kernel bug here, disk deletion code, fsync IO, checking for disk dead while queue is marked dying,
+	// which will never happen, and the disk deletion code is stuck in deadlock.
+	// in this case, allow IOs to go to the queue and let px-storage handle IO
+	// on the device, so fsync can pass, and del_gendisk can proceed
+	// to safely delete the device.
```

Comment:
Why not just have 2 cases only - one for < 5.13 like above and rest all in the else case?
No handling required.

Reply:
> blk_mark_disk_dead(pxd_dev->disk);

This fixed api is not available in between.

Reply:
I meant just assume that api is not available > 5.13. Use the solution we have for in between for all cases.
Anyways, no reason to hold the pr for that.